### PR TITLE
marshal.go: Fix signedness issue in marshalPDU()

### LINKED ISSUE #293: Unsigned to signed conversion error in SnmpPacket.marshalPDU()
SnmpPacket.MaxRepetitions is type uint8 throughout gosnmp.
https://github.com/gosnmp/gosnmp/blob/9f003f366c564e2bd597dd765f5b97cdcee821e3/marshal.go#L54

In SnmpPacket.marshalPDU(), MaxRepetitions is used to build a GetBulkRequest PDU where a signed value is expected.
https://github.com/gosnmp/gosnmp/blob/9f003f366c564e2bd597dd765f5b97cdcee821e3/marshal.go#L539

In this code, we're writing in ASN.1 BER (basic encoding rules).  The first byte is the data type, with 2 meaning signed integer, then the second byte is length of data in bytes, finally the data is one byte and comes from SnmpPacket.MaxRepetitions.

This unsigned to signed conversion error causes some agents (including net-snmp) to refuse to respond to the request.  Values 0 through 127 work fine in gosnmp because their signed representation is the same.  Unsigned values 128 through 255 don't fit in a signed 8 bit integer.  When they are misinterpreted as signed data they become -128 through -1.  Those negative values are invalid so an agent can't know how to handle them.

The SNMP RFC defines this PDU, the types of the data and their valid ranges.  Note that the INTEGER type here always means signed.
https://tools.ietf.org/html/rfc3416#section-3:
```
   max-bindings INTEGER ::= 2147483647
...
   BulkPDU ::=                         -- must be identical in
       SEQUENCE {                      -- structure to PDU
           request-id      INTEGER (-214783648..214783647),
           non-repeaters   INTEGER (0..max-bindings),
           max-repetitions INTEGER (0..max-bindings),

           variable-bindings           -- values are ignored
               VarBindList
       }
```

ASN.1 BER is defined in ITU-T X.690.  It's available at https://www.itu.int/ITU-T/studygroups/com17/languages/X.690-0207.pdf.  Section 8.3 describes how to encode an integer.

There are other places where gosnmp uses an unsigned integer in a BER context that expects a signed integer.  I'm not sure if these cause similar failures, but they should be checked.
https://github.com/gosnmp/gosnmp/blob/9f003f366c564e2bd597dd765f5b97cdcee821e3/marshal.go#L536
https://github.com/gosnmp/gosnmp/blob/9f003f366c564e2bd597dd765f5b97cdcee821e3/marshal.go#L563
https://github.com/gosnmp/gosnmp/blob/9f003f366c564e2bd597dd765f5b97cdcee821e3/marshal.go#L560

### PR DESCRIPTION
**Pull request by TimRots**
Fix for signedness issue in marshalPDU() causing
agents to refuse responding if MaxRepetitions was > 127.

Fixes #293

Signed-off-by: Tim Rots <tim.rots@protonmail.ch>

### CODE REVIEW

**Comment by SuperQ** [comment]
This seems like a breaking change to the public function interfaces. I don't think we have any public interface stability guarantees, but it would be good to note this in the CHANGELOG for users.

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,10 @@
 * [ENHANCEMENT]
 * [BUGFIX]
 
+NOTE:
+This release changes the ``MaxRepetitions`` type to uint32.
+
+* [BUGFIX] Fix GetBulkRequest MaxRepetitions signedness issue in marshalPDU() #293
```
Comment (SuperQ) [inline CHANGELOG.md:11]: Since this is breaking, let's label this `[CHANGE]`.

**Review by SuperQ — APPROVED** [review]
Minor nit on the changelog, otherwise LGTM.

**Comment by Hipska** [comment]
What is now the max allowed (working) value for Max Repetitions in gosnmp?

**Comment by TimRots** [comment]
> What is now the max allowed (working) value for Max Repetitions in gosnmp?

``2147483647``

**Review by reimda — COMMENTED** [review]
Thanks for the fix @TimRots!  Sorry not to be able to review this over the weekend before it was merged.  It prevents the negative values and covers the valid range so it looks good to me.

One thing that might be nice to add is documentation of the valid range and/or a runtime check that the values aren't out of range.  For example have SetMaxRepetitions() return an error if maxRepetitions is 0x7FFFFFFF or greater.

Without documentation or a runtime check the user might still get unexpected behavior.

**Comment by Hipska** [comment]
Any idea when a new release will come out?

**Comment by TimRots** [comment]
> Any idea when a new release will come out?

Sorry for only replying just now. We will do our best to release ``1.30.0`` this weekend

**Comment by TimRots** [comment]
> Thanks for the fix @TimRots! Sorry not to be able to review this over the weekend before it was merged. It prevents the negative values and covers the valid range so it looks good to me.

Very welcome @reimda. You're detailed report helped a lot to understand/fix this in a quickly manner :) Thank you for the patience to explain things clearly.

> One thing that might be nice to add is documentation of the valid range and/or a runtime check that the values aren't out of range. For example have SetMaxRepetitions() return an error if maxRepetitions is 0x7FFFFFFF or greater.
> 
> Without documentation or a runtime check the user might still get unexpected behavior.

+1 for documenting this. I will add this to my to-do list

**Pull request merged**